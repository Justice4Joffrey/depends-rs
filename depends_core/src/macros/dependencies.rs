use proc_macro2::{Ident, Span, TokenStream};
use quote::quote;
use syn::{parse2, spanned::Spanned, Field, Fields, Generics, ItemStruct};

/// In place until we figure out the dependencies generics story
fn block_generics(generics: &Generics) -> syn::Result<()> {
    if !generics.params.is_empty() {
        Err(syn::Error::new(
            generics.span(),
            "dependencies don't currently support generics",
        ))
    } else {
        Ok(())
    }
}

pub fn dependencies_attr(args: TokenStream, input: TokenStream) -> TokenStream {
    if !args.is_empty() {
        panic!("dependencies attribute does not support arguments")
    }
    let ItemStruct {
        vis,
        ident,
        generics,
        fields,
        ..
    } = parse2::<ItemStruct>(input).unwrap();

    // until we figure out how to resolve dependencies without propagating
    // throughout the graph, block
    block_generics(&generics).unwrap();
    let name = ident.to_string();
    let ref_ident = Ident::new(format!("{}Ref", name).as_str(), Span::call_site());
    let fields = if let Fields::Named(f) = fields {
        f.named
    } else {
        panic!("must be a struct with named fields");
    };

    let mut new_fields = TokenStream::new();
    let mut ref_fields = TokenStream::new();
    let mut field_args = TokenStream::new();
    let mut field_new_args = TokenStream::new();
    let mut field_resolves = TokenStream::new();
    let mut dirty_field_args = Vec::<TokenStream>::new();
    fields.into_iter().for_each(|Field { vis, ident, ty, .. }| {
        let ident = ident.expect("struct fields must be named.");
        let new_type = quote! { ::depends::core::Dependency<::std::sync::Arc<#ty>> };
        new_fields.extend(quote! {
            #vis #ident: #new_type,
        });
        ref_fields.extend(quote! {
            #vis #ident: <#new_type as ::depends::core::Resolve>::Output<'a>,
        });
        field_args.extend(quote! {
            #ident: ::std::sync::Arc<#ty>,
        });
        field_new_args.extend(quote! {
            #ident: ::depends::core::Dependency::new(#ident),
        });
        field_resolves.extend(quote! {
            #ident: self.#ident.resolve(visitor),
        });
        dirty_field_args.push(quote! {
            self.#ident.is_dirty()
        });
    });

    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();

    quote! {
        #vis struct #ident {
            #new_fields
        }

        #vis struct #ref_ident <'a> {
            #ref_fields
        }

        impl #impl_generics #ident #ty_generics #where_clause {
            pub fn new(#field_args) -> Self {
                Self {
                    #field_new_args
                }
            }
        }

        impl #impl_generics ::depends::core::Resolve for #ident #ty_generics #where_clause {
            type Output<'a> = #ref_ident<'a> where Self: 'a;

            fn resolve(&self, visitor: &mut impl ::depends::core::Visitor) -> Self::Output<'_> {
                #ref_ident {
                    #field_resolves
                }
            }
        }

        impl <'a> ::depends::core::IsDirty for #ref_ident<'a> {
            fn is_dirty(&self) -> bool {
                #(#dirty_field_args)||*
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use syn::parse_quote;

    use super::*;
    use crate::macros::helpers::format_source;

    #[test]
    #[ignore]
    fn test_dependencies() {
        let input = parse_quote! {
            struct Components {
                node1: Node1,
                node2: Node2,
                node3: Node3,
            }
        };

        assert_snapshot!(
            "dependencies",
            format_source(
                dependencies_attr(TokenStream::new(), input)
                    .to_string()
                    .as_str()
            )
        );
    }

    #[should_panic]
    #[test]
    fn test_dependencies_rejects_attrs() {
        let input = parse_quote! {
            struct Components {
                node1: Node1,
                node2: Node2,
                node3: Node3,
            }
        };

        dependencies_attr(quote! {thing = 1}, input);
    }
}
