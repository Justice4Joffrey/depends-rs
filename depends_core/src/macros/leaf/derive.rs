use proc_macro2::TokenStream;
use quote::quote;
use syn::{parse2, Data, DeriveInput, Field};

use super::parsed_attrs::LeafParsedAttrs;
use crate::macros::{
    leaf::LeafAttrModel,
    model::{get_depends_attrs, FieldAttrs},
    HashLogic,
};

pub fn derive_leaf(input: TokenStream) -> TokenStream {
    let DeriveInput {
        ident,
        data,
        generics,
        attrs,
        ..
    } = parse2::<DeriveInput>(input).unwrap();

    let data = if let Data::Struct(data) = data {
        data
    } else {
        panic!("This macro can only be derived for structs.");
    };

    let name = ident.to_string();
    let (custom_clean, hashing) = {
        let struct_attrs = get_depends_attrs(&attrs);
        let field_attrs: Vec<_> = data
            .fields
            .iter()
            .map(|Field { attrs, ident, .. }| {
                FieldAttrs {
                    ident: ident
                        .as_ref()
                        .expect("Only structs with named fields are supported.")
                        .clone(),
                    field_attrs: get_depends_attrs(attrs),
                }
            })
            .collect();

        let model = LeafAttrModel {
            struct_attrs,
            field_attrs,
        };
        let parsed: LeafParsedAttrs = model.try_into().unwrap();
        (
            parsed.custom_clean.unwrap_or(false),
            parsed.hashing.unwrap_or(HashLogic::Struct),
        )
    };

    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();

    let clean_clause = if custom_clean {
        TokenStream::new()
    } else {
        quote! {
            impl #ty_generics ::depends::core::Clean for #ident #ty_generics #where_clause {
                fn clean(&mut self) {}
            }
        }
    };

    let hash_value_clause = hashing.to_tokens();

    quote! {
        impl #impl_generics ::depends::core::Named for #ident #ty_generics #where_clause {
            fn name() -> &'static str {
                #name
            }
        }

        impl #impl_generics #ident #ty_generics #where_clause {
            pub fn into_leaf(self) -> ::std::sync::Arc<::depends::core::LeafNode<Self>> {
                ::depends::core::LeafNode::new(self)
            }
        }

        impl #impl_generics ::depends::core::HashValue for #ident #ty_generics #where_clause {
            fn hash_value(&self, hasher: &mut impl ::std::hash::Hasher) -> ::depends::core::NodeHash {
                #hash_value_clause
            }
        }

        #clean_clause
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use syn::parse_quote;

    use super::*;
    use crate::macros::helpers::format_source;

    #[test]
    #[ignore]
    fn test_leaf() {
        let input: TokenStream = parse_quote! {
            #[diesel(some_arg)]
            pub struct Foo {
                bar: Vec<usize>
            }
        };
        assert_snapshot!(
            "leaf",
            format_source(derive_leaf(input).to_string().as_str())
        );
    }

    #[test]
    #[ignore]
    fn test_leaf_generics() {
        let input = parse_quote! {
            struct Foo<T> {
                bar: Vec<usize>
            }
        };
        assert_snapshot!(
            "leaf_generics",
            format_source(derive_leaf(input).to_string().as_str())
        );
    }

    #[test]
    #[ignore]
    fn test_leaf_hashed() {
        let input = parse_quote! {
            struct Foo<T> {
                bar: Vec<usize>,
                #[depends(hash)]
                number: usize,
            }
        };
        assert_snapshot!(
            "leaf_hashed_attr",
            format_source(derive_leaf(input).to_string().as_str())
        );
    }

    #[test]
    #[ignore]
    fn test_leaf_unhashable() {
        let input = parse_quote! {
            #[depends(unhashable)]
            struct Foo<T> {
                bar: Vec<usize>,
                number: usize,
            }
        };
        assert_snapshot!(
            "leaf_unhashable",
            format_source(derive_leaf(input).to_string().as_str())
        );
    }

    #[test]
    #[ignore]
    fn test_leaf_generics_custom_clean() {
        let input = parse_quote! {
            #[depends(custom_clean)]
            struct Foo<T> {
                bar: Vec<usize>
            }
        };
        assert_snapshot!(
            "leaf_generics_custom_clean",
            format_source(derive_leaf(input).to_string().as_str())
        );
    }

    #[test]
    #[should_panic]
    fn test_leaf_multi_hash_attr_args() {
        let input = parse_quote! {
            #[depends(unhashable)]
            struct Foo<T> {
                bar: Vec<usize>,
                #[depends(hash)]
                number: usize,
            }
        };
        derive_leaf(input);
    }

    #[test]
    #[should_panic]
    fn test_leaf_multi_attrs() {
        let input = parse_quote! {
            struct Foo<T> {
                bar: Vec<usize>,
                #[depends(hash)]
                number: usize,
                #[depends(hash)]
                another: usize,
            }
        };
        derive_leaf(input);
    }

    #[test]
    #[should_panic]
    fn test_leaf_on_enum() {
        let input = parse_quote! {
            enum Foo {
                Thing(usize)
            }
        };

        derive_leaf(input);
    }
}
