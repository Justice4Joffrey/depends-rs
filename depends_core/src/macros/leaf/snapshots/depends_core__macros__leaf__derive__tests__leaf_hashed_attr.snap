---
source: depends_core/src/macros/leaf/derive.rs
assertion_line: 154
expression: format_source(derive_leaf(input).to_string().as_str())
---
impl<T> ::depends::core::Named for Foo<T> {
    fn name() -> &'static str {
        "Foo"
    }
}
impl<T> Foo<T> {
    pub fn into_leaf(self) -> ::std::sync::Arc<::depends::core::LeafNode<Self>> {
        ::depends::core::LeafNode::new(self)
    }
}
impl<T> ::depends::core::HashValue for Foo<T> {
    fn hash_value(&self, hasher: &mut impl ::std::hash::Hasher) -> ::depends::core::NodeHash {
        ::depends::core::NodeHash::Hashed({
            self.number.hash(hasher);
            hasher.finish()
        })
    }
}
impl<T> ::depends::core::Clean for Foo<T> {
    fn clean(&mut self) {}
}

